lib/philomena_web/plugs/limit_plug.ex

```diff
@@ -0,0 +1,74 @@
+defmodule PhilomenaWeb.LimitPlug do
+  @moduledoc """
+  This plug automatically limits requests which are submitted faster
+  than should be allowed for a given client.
+  ## Example
+      plug PhilomenaWeb.LimitPlug, [time: 30, error: "Too fast! Slow down."]
+  """
+
+  alias Plug.Conn
+  alias Phoenix.Controller
+  alias Philomena.Users.User
+
+  @doc false
+  @spec init(any()) :: any()
+  def init(opts), do: opts
+
+  @doc false
+  @spec call(Conn.t(), any()) :: Conn.t()
+  def call(conn, opts) do
+    limit = Keyword.get(opts, :limit, 1)
+    time = Keyword.get(opts, :time, 5)
+    error = Keyword.get(opts, :error)
+
+    data = [
+      current_user_id(conn.assigns.current_user),
+      :inet_parse.ntoa(conn.remote_ip),
+      conn.private.phoenix_action,
+      conn.private.phoenix_controller
+    ]
+
+    key = "rl-#{Enum.join(data, "")}"
+
+    [amt, _] =
+      Redix.pipeline!(:redix, [
+        ["INCR", key],
+        ["EXPIRE", key, time]
+      ])
+
+    cond do
+      amt <= limit ->
+        conn
+
+      is_staff(conn.assigns.current_user) ->
+        conn
+
+      ajax?(conn) ->
+        conn
+        |> Controller.put_flash(:error, error)
+        |> Conn.send_resp(:multiple_choices, "")
+        |> Conn.halt()
+
+      true ->
+        conn
+        |> Controller.put_flash(:error, error)
+        |> Controller.redirect(external: conn.assigns.referrer)
+        |> Conn.halt()
+    end
+  end
+
+  defp is_staff(%User{role: "admin"}), do: true
+  defp is_staff(%User{role: "moderator"}), do: true
+  defp is_staff(%User{role: "assistant"}), do: true
+  defp is_staff(_), do: false
```

Comment:
I feel like the staff bypass should canonically be done via `can?` and ability.ex, but it's not a strong feeling.

Reply:
would incremental improvement (with further PRs, whenever someone worked on this) be okay for you?

Reply:
It'll probably fall into a black hole and never be addressed, but that's fine.  Like I said, not a strong feeling.  :v

Reply:
Would be nice to actually implement error/debug logging into the app. Those both tasks could be resolve at the same time:
- change those limits into configuration options,
- make those error messages adjustable depending on configuration options,
- offer a simple logging API
- change hardcoded limits, permission-checks, and similar into easily adjustable options.